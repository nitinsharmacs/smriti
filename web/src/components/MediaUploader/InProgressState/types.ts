import type { MediaItemType } from 'src/components/MediaUploader/types';

export interface InProgressStateProps {
  achievedUploads: number;
  targetUploads: number;
  mediaItems: MediaItemType[];
  onStop?: () => void;
}
