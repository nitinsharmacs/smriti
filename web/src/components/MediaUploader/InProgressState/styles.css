.upload-progressbar {
  margin-top: 1.5em;
}

.items-completion-indicator {
  font-size: 0.8rem;
  color: var(--sub-text-fg-color);
  text-align: right;
}
