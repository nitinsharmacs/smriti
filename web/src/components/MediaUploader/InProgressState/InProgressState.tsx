import { useCallback, useState } from 'react';

import ProgressBar from 'src/components/ProgressBar/ProgressBar';

import Button from 'src/components/Button/Button';
import Modal from 'src/components/MediaUploader/Modal/Modal';
import CollapsableMediaItems from 'src/components/MediaUploader/CollapsableMediaItems/CollapsableMediaItems';

import './styles.css';
import type { InProgressStateProps } from './types';

const InProgressState = (props: InProgressStateProps) => {
  const [showMore, setShowMore] = useState<boolean>(false);

  const showMoreHandler = useCallback(() => {
    setShowMore((prev) => !prev);
  }, []);

  return (
    <Modal>
      <div className='upload-in-progress'>
        <h4>Media getting uploaded</h4>
        <div className='upload-progressbar'>
          <ProgressBar
            value={(props.achievedUploads / props.targetUploads) * 100}
            lineWeight={8}
          />
          <div className='items-completion-indicator'>
            {props.achievedUploads} of {props.targetUploads} done
          </div>
        </div>
        <div className='upload-controls'>
          <Button variant='contained' onClick={props.onStop}>
            Stop
          </Button>
          <Button
            variant='text'
            onClick={showMoreHandler}
            style={{ marginLeft: '0.5em' }}
          >
            {showMore ? 'Show less' : 'Show more'}
          </Button>
        </div>
        <CollapsableMediaItems mediaItems={props.mediaItems} open={showMore} />
      </div>
    </Modal>
  );
};

export default InProgressState;
