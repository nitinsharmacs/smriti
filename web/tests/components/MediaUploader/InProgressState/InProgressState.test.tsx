import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import InProgressState from 'src/components/MediaUploader/InProgressState/InProgressState';
import {
  MediaType,
  MediaUploadStatus,
} from 'src/components/MediaUploader/types';
import { describe, expect, it, vi } from 'vitest';

describe('InProgressState', () => {
  it('should render progress state', () => {
    const { container } = render(
      <InProgressState achievedUploads={0} targetUploads={2} mediaItems={[]} />
    );
    expect(container).toMatchSnapshot();
  });

  it('should show more/less', async () => {
    render(
      <InProgressState
        achievedUploads={0}
        targetUploads={2}
        mediaItems={[
          {
            id: '123',
            type: MediaType.Image,
            status: MediaUploadStatus.InProgress,
            progress: 12,
            name: 'image.png',
          },
          {
            id: '1234',
            type: MediaType.Video,
            status: MediaUploadStatus.InProgress,
            progress: 40,
            name: 'video.mp4',
          },
        ]}
      />
    );

    const showMoreBtn = screen.getAllByRole('button')[1];

    expect(screen.queryByText('image.png')).not.toBeInTheDocument();
    expect(screen.queryByText('video.mp4')).not.toBeInTheDocument();

    expect(showMoreBtn.textContent).toBe('Show more');

    await userEvent.click(showMoreBtn);

    expect(showMoreBtn.textContent).toBe('Show less');
    expect(screen.getByText('image.png')).toBeInTheDocument();
    expect(screen.getByText('video.mp4')).toBeInTheDocument();
  });

  it('should stop progress', async () => {
    const stopMock = vi.fn();

    render(
      <InProgressState
        achievedUploads={0}
        targetUploads={1}
        mediaItems={[
          {
            id: '123',
            type: MediaType.Image,
            status: MediaUploadStatus.InProgress,
            progress: 12,
            name: 'image.png',
          },
        ]}
        onStop={stopMock}
      />
    );

    const stopBtn = screen.getAllByRole('button')[0];

    await userEvent.click(stopBtn);

    expect(stopMock).toHaveBeenCalledOnce();
  });
});
