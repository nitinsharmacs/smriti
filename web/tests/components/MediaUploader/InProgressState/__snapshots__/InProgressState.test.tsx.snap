// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`InProgressState > should render progress state 1`] = `
<div>
  <div
    class="upload-txn-modal"
  >
    <div
      class="upload-in-progress"
    >
      <h4>
        Media getting uploaded
      </h4>
      <div
        class="upload-progressbar"
      >
        <span
          aria-valuemax="100"
          aria-valuemin="0"
          aria-valuenow="0"
          class="MuiLinearProgress-root MuiLinearProgress-colorPrimary MuiLinearProgress-determinate progressbar css-16mo9zg-MuiLinearProgress-root"
          role="progressbar"
        >
          <span
            class="MuiLinearProgress-bar MuiLinearProgress-bar1 MuiLinearProgress-barColorPrimary MuiLinearProgress-bar1Determinate css-l16vtb-MuiLinearProgress-bar1"
            style="transform: translateX(-100%);"
          />
        </span>
        <div
          class="items-completion-indicator"
        >
          0
           of 
          2
           done
        </div>
      </div>
      <div
        class="upload-controls"
      >
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary contained-btn css-74d805-MuiButtonBase-root-MuiButton-root"
          tabindex="0"
          type="button"
        >
          Stop
        </button>
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary text-btn css-1uent87-MuiButtonBase-root-MuiButton-root"
          style="margin-left: 0.5em;"
          tabindex="0"
          type="button"
        >
          Show more
        </button>
      </div>
      <div
        class="collapsable-media-items"
      />
    </div>
  </div>
</div>
`;
